### PR TITLE
fixes a big harddel source

### PR DESCRIPTION
**Pull request by Ben10Omintrix**
## About The Pull Request
this hard delete would happen everytime a player joins/observers the round. upon using the ref tracker i discovered this problem was caused by movable/screens not cleaning up after their owning huds if it gets deleted.  i didnt weakref it was cause it meant id have to resolve it in so many places, but if weakrefing is preferable i will do it
image

## Why It's Good For The Game
fixes a hard delete

### CODE REVIEW

**Comment by san7890** [comment]
weakrefs would be ideal if we didn't need hundreds of lines of boilerplate here

File: code/_onclick/hud/screen_objects.dm [hunk 1]
```diff
@@ -45,8 +45,10 @@
 
 /atom/movable/screen/Initialize(mapload, datum/hud/hud_owner)
 	. = ..()
-	if(hud_owner && istype(hud_owner))
-		hud = hud_owner
+	if(!istype(hud_owner))
+		return
+	hud = hud_owner
+	RegisterSignal(hud, COMSIG_QDELETING, PROC_REF(on_hud_delete))
```
Comment (LemonInTheDark) [inline code/_onclick/hud/screen_objects.dm:51]: this potentially breaks parallax.
Rather then doing something like this, create a setter proc that unreg's/registers QDELETING and such, and replace all other instances of modifying hud with it

Also, this is odd and is I think a patch fix.
Screen objects are meant to be owned by their hud, like the hud should manage and delete them and such.
What is the type HANGING the hud var?

**Comment by Ben10Omintrix** [comment]
should be ready for review now

**Comment by san7890** [comment]
@LemonInTheDark

**Review by Ghommie — APPROVED** [review]

File: code/_onclick/hud/screen_objects.dm [hunk 2]
```diff
@@ -45,8 +45,7 @@
 
 /atom/movable/screen/Initialize(mapload, datum/hud/hud_owner)
 	. = ..()
-	if(hud_owner && istype(hud_owner))
-		hud = hud_owner
```
Comment (LemonInTheDark) [inline code/_onclick/hud/screen_objects.dm:49]: is this the ONLY place we set hud in the codebase?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

**Comment by Ben10Omintrix** [comment]
sorry i missed the notification for these reviews

File: code/_onclick/hud/screen_objects.dm (hunk 2 above)
Comment (Ben10Omintrix) [reply code/_onclick/hud/screen_objects.dm:49]: from a glance, it seems so. since hud is also a private var

**Comment by LemonInTheDark** [comment]
no worry, all in good time

File: code/_onclick/hud/screen_objects.dm [hunk 3]
```diff
@@ -72,10 +71,24 @@
...
 	return
 
+///setter used to set our new hud
+/atom/movable/screen/proc/set_new_hud(datum/hud/hud_owner)
+	if(!istype(hud_owner))
+		return
+	if(hud)
+		UnregisterSignal(hud, COMSIG_QDELETING)
+	hud = hud_owner
+	RegisterSignal(hud, COMSIG_QDELETING, PROC_REF(on_hud_delete))
```
[the remark this answers is not in the document]
Comment (san7890) [reply code/_onclick/hud/screen_objects.dm:81]: i don't think it's a valid arg yet unless i'm misunderstanding

[the remark this answers is not in the document]
Comment (Ben10Omintrix) [reply code/_onclick/hud/screen_objects.dm:81]: whoops forgot to push changes

**Pull request merged**